crates/circuit/src/packed_instruction.rs

```diff
@@ -0,0 +1,497 @@
+///    bits of the full 64 bits are guaranteed to be zero (so one marked   +-- enumeration.  This
+///    `P` is always zero here), so we can retrieve the "full" pointer by      is 0b10, which means
+///    taking the whole `usize` and zeroing the low 3 bits, letting us         that this points to
+///    store the discriminant in there at other times.                         a `PyInstruction`.
+/// ```
+///
+/// There is currently one spare bit that could be used for additional metadata, if required.
+///
+/// # Construction
+///
+/// From Rust space, build this type using one of the `from_*` methods, depending on which
+/// implementer of `Operation` you have.  `StandardGate` has an implementation of `Into` for this.
+///
+/// From Python space, use the supplied `FromPyObject`.
+///
+/// # Safety
+///
+/// `PackedOperation` asserts ownership over its contained pointer (if not a `StandardGate`).  This
+/// has the following requirements:
+///
+/// * The pointer must be managed by a `Box` using the global allocator.
+/// * The pointed-to data must match the type of the discriminant used to store it.
+/// * `PackedOperation` must take care to forward implementations of `Clone` and `Drop` to the
+///   contained pointer.
+#[derive(Debug)]
+#[repr(transparent)]
+pub struct PackedOperation(usize);
+
+impl PackedOperation {
+    /// The bits representing the `PackedOperationType` discriminant.  This can be used to mask out
+    /// the discriminant, and defines the rest of the bit shifting.
+    const DISCRIMINANT_MASK: usize = 0b11;
+    /// The number of bits used to store the discriminant metadata.
+    const DISCRIMINANT_BITS: u32 = Self::DISCRIMINANT_MASK.count_ones();
+    /// A bitmask that masks out only the standard gate information.  This should always have the
+    /// same effect as `POINTER_MASK` because the high bits should be 0 for a `StandardGate`, but
+    /// this is defensive against us adding further metadata on `StandardGate` later.  After
+    /// masking, the resulting integer still needs shifting downwards to retrieve the standard gate.
+    const STANDARD_GATE_MASK: usize = (u8::MAX as usize) << Self::DISCRIMINANT_BITS;
+    /// A bitmask that retrieves the stored pointer directly.  The discriminant is stored in the
+    /// low pointer bits that are guaranteed to be 0 by alignment, so no shifting is required.
+    const POINTER_MASK: usize = usize::MAX ^ Self::DISCRIMINANT_MASK;
+
+    /// Extract the discriminant of the operation.
+    #[inline]
+    fn discriminant(&self) -> PackedOperationType {
+        ::bytemuck::checked::cast((self.0 & Self::DISCRIMINANT_MASK) as u8)
+    }
+
+    /// Get the contained pointer to the `PyGate`/`PyInstruction`/`PyOperation` that this object
+    /// contains.
+    ///
+    /// **Panics** if the object represents a standard gate; see `try_pointer`.
+    #[inline]
+    fn pointer(&self) -> NonNull<()> {
+        self.try_pointer().unwrap()
```

Comment:
I lean toward the "always use `expect` instead of `unwrap`" camp. So something like this
```suggestion
        self.try_pointer()
            .expect("Qiskit bug: PackedOperation contains a StandardGate, expecting a pointer")
```
All the logic is very local, so it's unlikely the code is ever changed in a way that triggers this bug. But it also helps the reader of the code understand what's going on. Might even replace the comment.

Reply:
This kind of message is guessing at what's gone wrong - we wouldn't _know_ that it includes a standard gate at this point, just that `try_pointer` failed for some reason.  Were we to expand `PackedOperation` in the future, such as by using the extra 4 discriminants we have available to us to encode something else (for example, maybe we allow something like a custom downstream C / Rust project to have its own `StandardGate` table as a fifth option), the error could easily become out-of-date as well.

This function is entirely private; a user can't violate the assumptions, only Qiskit can do that to itself. If we've done that, we can't really predict _how_ we violated them, and this error message wouldn't help with debugging it, it just makes a one-line function into a noisier two.

(I possibly should have done `unsafe fn pointer(&self) -> NonNull<()> { unsafe { self.try_pointer().unwrap_unchecked() } }` with this - this is a fast path for when it's _known_ safe.)

Reply:
I changed a couple of `unwrap`s to `expects` with a little note on the reason for the assumption, though I couldn't think of much more to say than the "panics" documentation section already attached.